# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/ThirdPartyCookieBlockingExceptions.h
# Commit: de6a26d3e3df
# Full Hash: de6a26d3e3dfd5d09121f11ae2b09903c780179b
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2024-12-17 17:00:49
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 4: Implement the third-party cookie blocking excpetion in CookieService. r=cookie-reviewers,baku,valentin
#   
#   The patch integrates the third-party cookie blocking excpetion in
#   CookieService.
#   
# ==============================================================================

diff -r b2934cae3c23 -r de6a26d3e3df netwerk/cookie/ThirdPartyCookieBlockingExceptions.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/netwerk/cookie/ThirdPartyCookieBlockingExceptions.h	Tue Dec 17 08:57:31 2024 +0000
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_net_ThirdPartyCookieBlockingExceptions_h
+#define mozilla_net_ThirdPartyCookieBlockingExceptions_h
+
+#include "mozilla/MozPromise.h"
+#include "nsEffectiveTLDService.h"
+#include "nsString.h"
+#include "nsTArray.h"
+#include "nsTHashSet.h"
+#include "nsIThirdPartyCookieBlockingExceptionListService.h"
+
+class nsIEffectiveTLDService;
+class nsIURI;
+class nsIChannel;
+
+namespace mozilla {
+namespace net {
+
+class ThirdPartyCookieBlockingExceptions final {
+ public:
+  // Lazily initializes the foreign cookie blocking exception list. The function
+  // returns the promise that resolves when the list is initialized.
+  RefPtr<GenericNonExclusivePromise> EnsureInitialized();
+
+  // Check if the given top-level and third-party URIs are in the exception
+  // list.
+  bool CheckExceptionForURIs(nsIURI* aFirstPartyURI, nsIURI* aThirdPartyURI);
+
+  // Check if the given channel is in the exception list.
+  bool CheckExceptionForChannel(nsIChannel* aChannel);
+
+  // Check if the given third-party site is in the wildcard exception list.
+  // The wildcard exception list is used to allow third-party cookies under
+  // every top-level site.
+  bool CheckWildcardException(const nsACString& aThirdPartySite);
+
+  // Check if the given first-party and third-party sites are in the exception
+  // list.
+  bool CheckException(const nsACString& aFirstPartySite,
+                      const nsACString& aThirdPartySite);
+
+  void Insert(const nsACString& aException);
+  void Remove(const nsACString& aException);
+
+  void GetExceptions(nsTArray<nsCString>& aExceptions);
+
+  void Shutdown();
+
+ private:
+  nsCOMPtr<nsIThirdPartyCookieBlockingExceptionListService>
+      m3PCBExceptionService;
+
+  // A helper function to create a key for the exception list.
+  static void Create3PCBExceptionKey(const nsACString& aFirstPartySite,
+                                     const nsACString& aThirdPartySite,
+                                     nsACString& aKey) {
+    aKey.Truncate();
+    aKey.Append(aFirstPartySite);
+    aKey.AppendLiteral(",");
+    aKey.Append(aThirdPartySite);
+  }
+
+  // The promise that resolves when the 3PCB exception service is initialized.
+  RefPtr<GenericNonExclusivePromise::Private> mInitPromise;
+  nsTHashSet<nsCStringHashKey> m3PCBExceptionsSet;
+};
+
+}  // namespace net
+}  // namespace mozilla
+
+#endif  // mozilla_net_ThirdPartyCookieBlockingExceptions_h